### ISSUE TITLE
"Extracting substitle tracks and attachments" takes too long 

### ISSUE THREAD

**Issue by reallyuniquename** [nixxquality/WebMConverter#98]
When I open 10-15GB MKV file (I tried different ones) for converting I get stuck at "Extracting substitle tracks and attachments" for more than 10-15 minutes while HDD constantly reads that file. Is it supposed to be like this?

**Comment by nixxquality** [comment]
If your mkv file contains multiple video or audio tracks, the program will pop up a dialog asking you to choose which one you would like to use in your output webm.

I suspect that you might have missed this and it's hidden somewhere. Search your desktop!

If that's not it, please respond again.

**Comment by reallyuniquename** [comment]
Yes, there is a dialog but it pops up after "Indexing..." part (which lasts about 1-2 minutes) and before "Extracting [...]" part. 
I select Audio 1 (for example) and then I have to wait like 10-15 minutes staring at "Extracting subtitle tracks and attachments" progress bar until I can proceed to converting.

**Comment by nixxquality** [comment]
I guess that can happen if the video file has a lot of attachments. I've never thought it would take up to 15 minutes, though.

I need to think about what I can do to fix this.

**Comment by reallyuniquename** [comment]
Ok, thanks. MKVs I'm trying to convert usually have 2 audio tracks and 2-3 subtitles though.

**Comment by nixxquality** [comment]
If you want to work around this issue manually, follow the [instructions I've laid out on the wiki](https://github.com/nixxquality/WebMConverter/wiki/Circumventing-long-indexing-times-with-mkv-splitting).

**Comment by reallyuniquename** [comment]
So basically you suggest to split MKV using another tool before converting video?

**Comment by nixxquality** [comment]
Yes. It won't mess with the encoded video/audio data at all, it should just give you a smaller slice to work with.

**Comment by reallyuniquename** [comment]
Oh ok. I'm gonna wait for the proper fix though.

**Comment by nixxquality** [comment]
I just released [v2.14.0](https://github.com/nixxquality/WebMConverter/releases/tag/v2.14.0).

It will skip any attachments that are already extracted, and you will be able to see what's taking so much time by using the Details button.

Please tell me if this alleviates your issue.

**Comment by reallyuniquename** [comment]
Ok, I've tried new build.

> Found subtitle track #5
> Extracting...

It takes 5-6 minutes to extract **one** subtitle track from MKV file. So demuxing MKV containing 4-5 subtitle tracks can take up to half an hour. Not sure why this happens. 
I tried mkvextract and it seems it takes same amount of time. I guess it's not a WebMConverter related issue then. 
It would be nice if someone else can confirm this.

**Comment by Zerathos** [comment]
I just tried it on a file in the latest version and it took less than 5 seconds to go through the extraction. I have noticed using mkvextract before that some files take longer than others though when I try to extract things out of them. If I had to guess at a reason, I'd say it's searching sequentially through the file's contents and the requested information isn't until the end of the file, which could take a while if the file is large.

**Comment by nixxquality** [comment]
Your varying extraction times is because you are extracting different types of extra data.
While attachments are pretty simple, a single block of data that's really just another file, subtitles are spread out in the same packets as audio/video data across the entire file.
It should only extract the subtitles once per input file though.

I guess the alternative would be moving the subtitle extraction process to when you want to add the Subtitle filter. I'll think about it.
